Documentation/guides/D8andR8.md

```diff
@@ -0,0 +1,270 @@
+
+But for an existing project, a developer could opt-in to the new
+behavior with two properties:
+```xml
+<Project>
+    <PropertyGroup>
+        <AndroidEnableProguard>True</AndroidEnableProguard>
+        <AndroidEnableMultiDex>True</AndroidEnableMultiDex>
+        <AndroidEnableDesugar>True</AndroidEnableDesugar>
+        <!--New properties-->
+        <AndroidDexTool>d8</AndroidDexTool>
+        <AndroidLinkTool>r8</AndroidLinkTool>
+    </PropertyGroup>
+</Project>
+```
+
+There should be two new MSBuild properties to configure here, because:
+- You could use `D8` in combination with `proguard`, as `R8` is not
+  "feature complete" in comparison to `proguard`.
+- You may not want to use code shrinking at all, but still use `D8`
+  instead of `dx`.
+- You shouldn't be able to use `dx` in combination with `R8`, it
+  doesn't make sense.
+- Developers should be able to use the existing properties for
+  enabling code shrinking, `multidex`, and `desugar`.
+
+Our reasonable defaults would be:
+- If `AndroidDexTool` is omitted, `dx` and `CompileToDalvik`
+  should be used. Until D8/R8 integration is deemed stable and enabled
+  by default.
+- If `AndroidDexTool` is `d8` and `AndroidEnableDesugar` is
+  omitted, `AndroidEnableDesugar` should be enabled.
+- If `AndroidLinkTool` is omitted and `AndroidEnableProguard` is
+  `true`, we should default `AndroidLinkTool` to `proguard`.
+
+MSBuild properties default to:
+```xml
+<AndroidDexTool Condition=" '$(AndroidDexTool)' == '' ">dx</AndroidDexTool>
+<!--NOTE: $(AndroidLinkTool) would be blank if code shrinking is not used at all-->
+<AndroidLinkTool Condition=" '$(AndroidLinkTool)' == '' And '$(AndroidEnableProguard)' == 'True' ">proguard</AndroidLinkTool>
+<AndroidEnableDesugar Condition=" '$(AndroidEnableDesugar)' == '' And ('$(AndroidDexTool)' == 'd8' Or '$(AndroidLinkTool)' == 'r8') ">True</AndroidEnableDesugar>
+```
+
+If a user specifies combinations of properties:
+- `AndroidDexTool` = `d8` and `AndroidEnableProguard` = `True`
+  - `AndroidLinkTool` will get set to `proguard`
+- `AndroidDexTool` = `dx` and `AndroidLinkTool` = `r8`
+  - This combination doesn't really *make sense*, but we don't need to
+    do anything: only `R8` will be called because it dexes and shrinks
+    at the same time.
+- `AndroidEnableDesugar` is enabled when omitted, if either `d8` or
+  `r8` are used
+
+For new projects that want to use D8/R8, code shrinking, and
+`multidex`, it would make sense to specify:
+```xml
+<Project>
+    <PropertyGroup>
+        <AndroidEnableMultiDex>True</AndroidEnableMultiDex>
+        <AndroidDexTool>d8</AndroidDexTool>
+        <AndroidLinkTool>r8</AndroidLinkTool>
+    </PropertyGroup>
+</Project>
+```
+
+# Additional D8 / R8 settings?
+
+`--debug` or `--release` needs to be explicitly specified for both D8
+and R8. We use the [AndroidIncludeDebugSymbols][debug_symbols]
+property for this.
+
+`$(D8ExtraArguments)` and `$(R8ExtraArguments)` can be used to
```

Comment:
Should we be prefixing these with `Android`? eg. `AndroidD8ExtraArguments` ?

Reply:
👍 whoops, yeah this will be public, so should be prefixed with `Android`.